File: Course Schedule - Leetcode 207/Course Schedule - Leetcode 207.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> graph;
        for (const auto& pair : prerequisites) {
            graph[pair[0]].push_back(pair[1]);
        }

        vector<int> states(numCourses, 0); // 0: UNVISITED, 1: VISITING, 2: VISITED

        for (int i = 0; i < numCourses; ++i) {
            if (!dfs(i, graph, states)) {
                return false;
            }
        }

        return true;
    }

private:
    bool dfs(int node, unordered_map<int, vector<int>>& graph, vector<int>& states) {
        if (states[node] == 1) return false; // Cycle detected
        if (states[node] == 2) return true;  // Already visited

        states[node] = 1; // Mark as visiting

        for (int neighbor : graph[node]) {
            if (!dfs(neighbor, graph, states)) {
                return false;
            }
        }

        states[node] = 2; // Mark as visited
        return true;
    }
};

File: Course Schedule - Leetcode 207/Course Schedule - Leetcode 207.java
import java.util.*;

public class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] pair : prerequisites) {
            graph.computeIfAbsent(pair[0], k -> new ArrayList<>()).add(pair[1]);
        }

        final int UNVISITED = 0;
        final int VISITING = 1;
        final int VISITED = 2;
        int[] states = new int[numCourses];

        for (int i = 0; i < numCourses; i++) {
            if (!dfs(i, graph, states)) {
                return false;
            }
        }

        return true;
    }

    private boolean dfs(int node, Map<Integer, List<Integer>> graph, int[] states) {
        if (states[node] == 1) return false; // Cycle detected
        if (states[node] == 2) return true;  // Already visited

        states[node] = 1; // Mark as visiting

        if (graph.containsKey(node)) {
            for (int neighbor : graph.get(node)) {
                if (!dfs(neighbor, graph, states)) {
                    return false;
                }
            }
        }

        states[node] = 2; // Mark as visited
        return true;
    }
}
